Synthetic wavelength calibration data

tepickering opened the pull request:
This PR adds a utility for creating realistic models of 2D spectroscopic images of arc lamp (or sky background) calibration observations. This will be key for testing the various stages of the wavelength calibration code and process as it's developed. Tilts/curvature along the spatial axis is modeled using 1D polynomials. Non-linear dispersion relations are modeled using the ideal disperser model implemented within the FITS WCS standard. Examples of the use of this routine are included in its docstrings. 

A few other notes:

- Use of the `specreduce_data` dependency has been deprecated. Cleaner to just always download/cache the data as needed. 
- https://github.com/pypeit/PypeIt is used as the source for comparison line lists because it has a large selection of well-vetted ones. 
- In the example where a non-linear dispersion relation is created using the WCS ideal disperser model, I create the model by putting together the appropriate FITS header manually and then loading that into `astropy.wcs.WCS`. It would be useful to have a documented functional interface for building WCS spectrograph models that provides meaningful parameter names and units handling. However, I think that is outside of the scope of this PR. It's also an open question whether such a utility should live here or within `specutils`.

kecnry commented on specreduce/utils/synth_data.py:
```diff
@@ -68,3 +77,234 @@ def make_2dspec_image(
+    wave_air : bool (default: False)
+        If True, convert the vacuum wavelengths used by ``pypeit`` to air wavelengths.
+    wave_unit : `~astropy.units.Quantity`
+        If ``wcs`` is not provides, this defines the wavelength units of the
```
```suggestion
        If ``wcs`` is not provided, this defines the wavelength units of the
```

tepickering replied:
gah, fat fingers...

kecnry commented on specreduce/utils/synth_data.py:
```diff
@@ -68,3 +77,234 @@ def make_2dspec_image(
+    linelists : str or list of str (default: ['HeI'])
+        Specification for linelists to load from ``pypeit``
+    amplitude_scale : float (default: 1)
+        Scale factor to apply to amplitudes provides in the linelists
```
```suggestion
        Scale factor to apply to amplitudes provided in the linelists
```

kecnry commented on specreduce/utils/synth_data.py:
```diff
@@ -68,3 +77,234 @@ def make_2dspec_image(
+        of the dispersion axis.
+    wave_air : bool (default: False)
+        If True, convert the vacuum wavelengths used by ``pypeit`` to air wavelengths.
+    wave_unit : `~astropy.units.Quantity`
```
should this be a `Quantity` or `Unit` object?

tepickering replied:
ah ha! good point. it should in fact be `Unit`...

kecnry commented on specreduce/utils/synth_data.py:
```diff
@@ -68,3 +77,234 @@ def make_2dspec_image(
+        wcs = WCS(naxis=2)
+        wcs.wcs.ctype[0] = 'WAVE'
+        wcs.wcs.ctype[1] = 'PIXEL'
+        wcs.wcs.cunit[0] = wave_unit
```
do we need to check if the unit is valid (`u.Unit` and spectral)?

kecnry commented on specreduce/utils/synth_data.py:
```diff
@@ -68,3 +77,234 @@ def make_2dspec_image(
+    extent : 2-element list-like
+        If ``wcs`` is not provided, this defines the beginning and end wavelengths
+        of the dispersion axis.
+    wave_air : bool (default: False)
+        If True, convert the vacuum wavelengths used by ``pypeit`` to air wavelengths.
```
having this between the two arguments that are only applicable if `wcs` is not provided might be confusing - could we move it below `extent` and `wave_unit`?

tepickering replied:
i'll at least do that if there's not a more clever way around it...

kecnry reviewed (approved):
Just a few minor comments/suggestions, but otherwise the code and tests look reasonable to me.  This will definitely be useful to use as we write more tests for wavelength calibration and spectral extraction.

tepickering commented on specreduce/utils/synth_data.py:
```diff
@@ -7,6 +7,15 @@
+from specreduce.calibration_data import load_pypeit_calibration_lines
+
+__all__ = [
+    'make_2dspec_image',
+    'make_2d_arc_image'
```
how about renaming `make_2dspec_image` to `make_2d_trace_image`? that's a more accurate description of what it does. this would open up the possibility of a `make_2d_spec_image` that combines `make_2d_trace_image` and `make_2d_arc_image`. might be worth implementing that as part of this PR.

tepickering commented:
thanks for the review, @kecnry! i think i addressed all of the comments. i also added a test i forgot to earlier and added a few more checks to flesh out coverage. if there are no other comments/concerns, i'll go ahead and merge in the next day or two.

pllim commented on specreduce/calibration_data.py:
```diff
@@ -62,28 +52,81 @@
 
-def get_reference_file_path(path=None, cache=False, show_progress=False):
+
+def get_reference_file_path(
```
Instead of rolling your own here, why not use the stuff in `astropy.utils.data` but use your own [conf.dataurl](https://github.com/astropy/astropy/blob/b40081ff722c5ca0a4ff1578abe4587e44c0eeba/astropy/utils/data.py#L99)?

https://github.com/astropy/astropy/blob/main/astropy/utils/data.py

tepickering replied:
when this was originally implemented years ago, i followed examples in https://docs.astropy.org/en/stable/utils/data.html and just wrapped `astropy.utils.data.download_file()` with some error handling. it looks like those examples haven't changed much since. i'll look into refactoring this to use pure-`astropy` at a higher level. that's out of scope for this PR, though.

tepickering replied:
i opened #169 to cover this. it's a little unclear to me how best to handle multiple data URLs with the `conf.dataurl` scheme. will play around more when i get the chance.

pllim commented on specreduce/tests/test_get_reference_file_path.py:
```diff
@@ -8,3 +8,12 @@ def test_get_reference_file_path():
     assert p is not None
+
+
+def test_get_pypeit_data_path():
```
If your test downloads data, you might wanna mark it with `@pytest.mark.remote_data`.

tepickering replied:
good call. i'll add that.

tepickering replied:
done. i tweaked the CI workflow so that `--remote-data` is only enabled for the one test where coverage is measured. this should hopefully help with the occasional failures due to downloads timing out. no longer parallel tests getting the same files at the same time...

tepickering commented on specreduce/calibration_data.py:
```diff
@@ -62,28 +52,81 @@
     """
-    Basic function to take a path to a file and load it via ``pkg_resources`` if
-    the ``specreduce_data`` package is available and load it via GitHub raw user content if not.
+    Utility to load reference data via GitHub raw user content. By default the ``specreduce_data``
```
the use of the package is deprecated. the repository is still there and will be there unless and until something better comes along.

pllim replied:
OK, I opened https://github.com/astropy/specreduce-data/issues/13

pllim commented:
Mine was just a drive-by review and not thorough. I'll leave it to specreduce maintainers to do the rest. Thanks!

tepickering commented:
drive-by or not, it was useful so thanks again, @pllim!
